### PR TITLE
[FLINK-5303] [table] Support for SQL GROUPING SETS clause.

### PR DESCRIPTION
**Pull request by chermenin**
Support for operators GROUPING SETS / ROLLUP / CUBE was added in this PR.
Also added some tests for check execution of SQL queries with them.
PR will close next issue: https://issues.apache.org/jira/browse/FLINK-5303.

### CODE REVIEW

**Comment by KurtYoung** [comment]
hi @chermenin , can you try to rebase this first, thanks.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetAggregate.scala [hunk 1]
```diff
@@ -104,7 +107,9 @@ class DataSetAggregate(
       namedAggregates,
       inputType,
       rowRelDataType,
-      grouping)
+      grouping,
+      indicator
+    )
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetAggregate.scala]: No need to have new line here.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamAggregate.scala [hunk 2]
```diff
@@ -70,7 +71,9 @@ class DataStreamAggregate(
       namedAggregates,
       getRowType,
       inputType,
-      grouping)
+      grouping,
+      indicator
+    )
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamAggregate.scala]: No need to have new line here.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamAggregate.scala [hunk 3]
```diff
@@ -207,7 +210,9 @@ class DataStreamAggregate(
             inputType,
             rowRelDataType,
             grouping,
-            namedProperties)
+            indicator,
+            namedProperties
+          )
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamAggregate.scala]: Same as here and a couple of other places

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateReduceCombineFunction.scala [hunk 4]
```diff
@@ -41,12 +41,14 @@ class AggregateReduceCombineFunction(
     private val aggregates: Array[Aggregate[_ <: Any]],
     private val groupKeysMapping: Array[(Int, Int)],
     private val aggregateMapping: Array[(Int, Int)],
+    private val additionalMapping: Array[(Int, Int)],
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateReduceCombineFunction.scala]: Can we have a more explicit name about "what is the additional mapping?". And we should update the class document.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateReduceGroupFunction.scala [hunk 5]
```diff
@@ -40,6 +40,7 @@ class AggregateReduceGroupFunction(
     private val aggregates: Array[Aggregate[_ <: Any]],
     private val groupKeysMapping: Array[(Int, Int)],
     private val aggregateMapping: Array[(Int, Int)],
+    private val additionalMapping: Array[(Int, Int)],
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateReduceGroupFunction.scala]: Please update the class document

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala [hunk 6]
```diff
@@ -98,7 +98,9 @@ object AggregateUtil {
     namedAggregates: Seq[CalcitePair[AggregateCall, String]],
     inputType: RelDataType,
     outputType: RelDataType,
-    groupings: Array[Int]): RichGroupReduceFunction[Row, Row] = {
+    groupings: Array[Int],
+    indicator: Boolean
```
Comment (KurtYoung) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala]: This variable is not used

**Comment by chermenin** [comment]
I cleaned up my code and I will write tests for the plan at this weekend.

**Comment by twalthr** [comment]
I would like to shepherd this PR.

File: flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/batch/sql/GroupingSetsTest.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+				"null,Comment#6,12,2\nnull,Comment#5,11,2\nnull,Comment#4,10,2\n" +
+				"null,Comment#3,9,2\nnull,Comment#2,8,2\nnull,Comment#15,21,2\n" +
+				"null,Comment#14,20,2\nnull,Comment#13,19,2\nnull,Comment#12,18,2\n" +
+				"null,Comment#11,17,2\nnull,Comment#10,16,2\nnull,Comment#1,7,2";
+
+		checkSql(query, expected);
+	}
+
+	@Test
+	public void testGroupingSetsWithNulls() throws Exception {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/java/org/apache/flink/table/api/java/batch/sql/GroupingSetsTest.java]: Can you add `GROUPING()` and `GROUPING_ID()` here too. Just to make sure that we don't compare two wrong results in the tests below.

**Comment by chermenin** [comment]
Big thanks for the review @twalthr! I fixed the code for most of the comments. I will fix the tests at this weekend.

**Comment by twalthr** [comment]
Thanks for the update @chermenin. I will have a final look over the code and merge it. I will also add a follow-up issue for efficiency.

**Pull request closed without merging**